File: leetsync/1387-find-elements-in-a-contaminated-binary-tree/find-elements-in-a-contaminated-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 * int val;
 * TreeNode left;
 * TreeNode right;
 * TreeNode() {}
 * TreeNode(int val) { this.val = val; }
 * TreeNode(int val, TreeNode left, TreeNode right) {
 * this.val = val;
 * this.left = left;
 * this.right = right;
 * }
 * }
 */
class FindElements {
    Set<Integer> set;

    public FindElements(TreeNode root) {
        set = new HashSet<>();
        bfs(root, 0);
    }

    public void bfs(TreeNode root, int value) {
        Queue<TreeNode> q = new LinkedList<>();
        root.val = value;
        q.add(root);
        while (!q.isEmpty()) {
            TreeNode current = q.poll();
            set.add(current.val);
            if (current.left != null) {
                current.left.val = current.val * 2 + 1;
                q.add(current.left);
            }
            if (current.right != null) {
                current.right.val = current.val * 2 + 2;
                q.add(current.right);
            }
        }
    }

    public boolean find(int target) {
        return set.contains(target);
    }
}

/**
 * Your FindElements object will be instantiated and called as such:
 * FindElements obj = new FindElements(root);
 * boolean param_1 = obj.find(target);
 */

File: leetsync/1387-find-elements-in-a-contaminated-binary-tree/find-elements-in-a-contaminated-binary-tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
type FindElements struct {
	set map[int]bool
}

func Constructor(root *TreeNode) FindElements {
	set := make(map[int]bool)
	dfs(root, 0, set)
	return FindElements{
		set: set,
	}
}

func dfs(root *TreeNode, value int, set map[int]bool) {
	if root == nil {
		return
	}
	root.Val = value
	set[value] = true
	dfs(root.Left, value*2+1, set)
	dfs(root.Right, value*2+2, set)
}

func (this *FindElements) Find(target int) bool {
	return this.set[target]
}

/**
 * Your FindElements object will be instantiated and called as such:
 * obj := Constructor(root);
 * param_1 := obj.Find(target);
 */